libs/checkpoint-duckdb/langgraph/store/duckdb/base.py

```diff
@@ -0,0 +1,395 @@
+import asyncio
+import json
+import logging
+from collections import defaultdict
+from contextlib import contextmanager
+from typing import (
+    Any,
+    Generic,
+    Iterable,
+    Iterator,
+    Sequence,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import duckdb
+from langgraph.store.base import (
+    BaseStore,
+    GetOp,
+    Item,
+    ListNamespacesOp,
+    Op,
+    PutOp,
+    Result,
+    SearchOp,
+)
+
+logger = logging.getLogger(__name__)
+
+
+MIGRATIONS = [
+    """
+CREATE TABLE IF NOT EXISTS store (
+    prefix TEXT NOT NULL,
+    key TEXT NOT NULL,
+    value JSON NOT NULL,
+    created_at TIMESTAMP DEFAULT now(),
+    updated_at TIMESTAMP DEFAULT now(),
+    PRIMARY KEY (prefix, key)
+);
+""",
+    """
+CREATE INDEX IF NOT EXISTS store_prefix_idx ON store (prefix);
+""",
+]
+
+C = TypeVar("C", bound=duckdb.DuckDBPyConnection)
+
+
+class BaseDuckDBStore(Generic[C]):
+    MIGRATIONS = MIGRATIONS
+    conn: C
+
+    def _get_batch_GET_ops_queries(
+        self,
+        get_ops: Sequence[tuple[int, GetOp]],
+    ) -> list[tuple[str, tuple, tuple[str, ...], list]]:
+        namespace_groups = defaultdict(list)
+        for idx, op in get_ops:
+            namespace_groups[op.namespace].append((idx, op.key))
+        results = []
+        for namespace, items in namespace_groups.items():
+            _, keys = zip(*items)
+            keys_to_query = ",".join(["?"] * len(keys))
+            query = f"""
+                SELECT prefix, key, value, created_at, updated_at
+                FROM store
+                WHERE prefix = ? AND key IN ({keys_to_query})
+            """
+            params = (_namespace_to_text(namespace), *keys)
+            results.append((query, params, namespace, items))
+        return results
+
+    def _get_batch_PUT_queries(
+        self,
+        put_ops: Sequence[tuple[int, PutOp]],
+    ) -> list[tuple[str, Sequence]]:
+        inserts: list[PutOp] = []
+        deletes: list[PutOp] = []
+        for _, op in put_ops:
```

Comment:
I see this was copied over from the Postgres implementation.

I'm wondering if there's a scenario where breaking the ordering of `DELETE` and `INSERT` in the original list of `put_ops` causes an issue.

This method accepts an arbitrary list of `put_ops` and returns a reordered list of queries.

Reply:
i think this is a valid concern, kind of related to Eugene's questions above. i think we'll need to rethink this a bit for both postgres and duckdb, but for now all of the publicly documented methods that we show in the docs are single ops https://github.com/langchain-ai/langgraph/blob/main/libs/checkpoint/langgraph/store/base/__init__.py#L260